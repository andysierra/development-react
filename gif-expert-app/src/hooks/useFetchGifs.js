import {useState, useEffect} from 'react';
import { giphySearch } from '../helpers/Giphy';

export const useFetchGifs = (category, limit=10) => {

  const [state, setState] = useState({
    loading:  true
  });

  useEffect(()=>{

    giphySearch(category, limit)
      .then(
        response => setState({...response, loading: false})
      );

  }, [category, limit]);
  
  return state;
}