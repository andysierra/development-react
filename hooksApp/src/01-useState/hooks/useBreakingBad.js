import PropTypes from 'prop-types';

export const useBreakingBad = () => {

  const url = 'https://www.breakingbadapi.com/api/quotes/';

  async function getQuote(id=1) {
    if(typeof id === 'number') {
      const received = await fetch(url+id);

      return received.json();
    }
    else console.log("getCuote has received an invalid number: "+typeof(id));
    return undefined;
  }

  return getQuote;
}