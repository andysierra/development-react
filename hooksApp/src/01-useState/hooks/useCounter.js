import PropTypes from 'prop-types';
import { useState } from 'react';

export const useCounter = ( defaultValue = 0 ) => {
  
  const [counter, setCounter] = useState(defaultValue);

  const increment = (step=1) => {
    if(typeof(step) !== 'number') step = 1;
    const newVal = counter+step;
    setCounter(newVal);
    return newVal;
  }
  const reset = () => {
    setCounter(defaultValue);
    return defaultValue;
  }
  const decrement = (step=1) => {
    if(typeof(step) !== 'number') step = 1;
    if(counter < 1) return;
    const newVal = counter-step;
    setCounter(newVal);
    return newVal;
  }

  return {
    counter,
    increment,
    reset,
    decrement
  }
}

useCounter.propTypes = {
  defaultValue: PropTypes.number
}