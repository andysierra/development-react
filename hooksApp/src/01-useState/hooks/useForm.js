import { useState } from "react";

export const useForm = ( defaultForm = {} ) => {
  
  const [form, setForm] = useState(defaultForm);

  const onInputChange = ({target}) => {
    const {name, value} = target;

    // ... validation stuff

    setForm({
      ...form,
      [name]: value
    });
  }

  const onReset = () => { setForm(defaultForm) }

  return {
    ...form,
    form,
    onInputChange,
    onReset
  }
}