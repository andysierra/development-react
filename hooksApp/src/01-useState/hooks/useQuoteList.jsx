import React, { useState } from "react";
import { Quote } from "../components/Quote";

export const useQuoteList = (sortAlign = 'center', justify='start', initial = []) => {

  const [list, setList] = useState(initial);

  function add(promise) {
    let   newList = [];
    const id = list.length.toString()+'_'+Date.now();
    const newQuote = 
      <Quote
        key     = {id}
        promise = {promise}
        align   = {sortAlign === 'stairs' ? (list.length % 3) : sortAlign}>
      </Quote>;

    switch(justify) {
      case 'end':   newList = [...list, newQuote]; break;
      default:      newList = [newQuote, ...list]; break;
    }

    setList(newList);
  }

  return ({ list, add });
}
