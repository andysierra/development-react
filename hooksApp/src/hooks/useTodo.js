import { todoReducer } from "../08-useReducer/TodoReducer";
import { useReducer } from "react";

const initialTodo = [
  {
    id:           'todoId_0_'+Date.now(),
    description:  'Ir a la iglesia',
    done:         false
  },
  {
    id:           'todoId_1_'+Date.now(),
    description:  'Ir a hacer mercado',
    done:         true
  },
  {
    id:           'todoId_2_'+Date.now(),
    description:  'Mover el bote en la disco',
    done:         false
  }
];

// actions:
const doneByIdAction    = { type: "marks as done a todo by id" };
const deleteByIdAction  = { type: "delete a todo by id" };

export const useTodo = () => {
  
  const [todos, dispatch] = useReducer(todoReducer, initialTodo);
  
  const doneById    = (id) => dispatch({...doneByIdAction, payload: id});
  const deleteById  = (id) => dispatch({...deleteByIdAction, payload: id});

  return {
    todos,
    doneById,
    deleteById
  }
}
