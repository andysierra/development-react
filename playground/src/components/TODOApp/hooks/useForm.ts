import { ChangeEvent, useEffect, useState } from "react"

export const useForm = (initialForm:object = {}) => {
  
  const [form, setForm]   = useState<object>(initialForm);
  const [ready, setReady] = useState<boolean>(true);  

  useEffect(() => {
    setReady(true);
  }, [form])

  const onChange = (e:ChangeEvent<HTMLInputElement> | ChangeEvent<HTMLTextAreaElement>) => {
    const { name, value } = e.target;
    setForm( prev => ({...prev, [name]: value}) );
    setReady(false);
  }

  const clean = () => {
    setForm(initialForm);
  }

  return [form, onChange, clean, ready] as const;
}
