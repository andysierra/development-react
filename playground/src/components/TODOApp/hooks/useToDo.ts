import React, { useReducer } from 'react'
import { initialState, reducer } from '../reducers/reducer';
import { ToDo } from '../types/toDoTypes';

export const useToDo = () => {

  // En este TODO app sencillo, podemos usar un simple reducer,
  // pero en casos de mÃ¡s profundidad de componente, mejor usar context
  const [state, dispatch] = useReducer(reducer, initialState);

  const deleteToDo = (id:number) : ToDo => {
    
  }

  return [
    state,
    dispatch,
    deleteToDo
  ] as const;
}
