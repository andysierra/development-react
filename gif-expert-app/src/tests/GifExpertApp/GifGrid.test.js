import { shallow } from "enzyme";
import { useFetchGifs } from "../../hooks/useFetchGifs";
import GifGrid from "../../pages/GifExpertApp/GifGrid";

// Mockeo el custom hook:
jest.mock("../../hooks/useFetchGifs");

describe('pruebas sobre Gifgrid.js', () => {

  let category = "pokemon";
  let wrapper = shallow(<GifGrid cat={category} />);

  beforeEach(()=>{
    wrapper = shallow(<GifGrid cat={category} />);
  });

  test('debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar items cuando se cargan las imagenes con useFetchGifs', () => {
    
  });
})