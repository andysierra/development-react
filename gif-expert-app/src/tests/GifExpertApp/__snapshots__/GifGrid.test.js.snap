// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas sobre Gifgrid.js debe mostrarse correctamente 1`] = `
<div
  className="card my-3 shadow-sm fade-in-down"
>
  <div
    className="card-header"
  >
    <div
      className="d-flex flex-row justify-content-between card-header-content"
    >
      <div
        className="d-flex flex-row flex-nowrap"
      >
        <h3>
          pokemon
        </h3>
      </div>
      <h5>
         
        Esperando 
         resultados 
      </h5>
    </div>
  </div>
  <div
    className="card-body"
  >
    <div
      className="row"
    >
      <GifGridItemLoading
        placeholders={10}
      />
    </div>
  </div>
</div>
`;
